src/set_relation/set_relation.cc

```diff
@@ -1642,6 +1642,225 @@ Set* Conjunction::normalize() const {
     return retval2;  
 }
 
+//*****************************************************************************
+
+bool Conjunction::isUFSArg(int tvar)
+{
+    bool is = false;
+    // Iterate over the equality expressions.
+    for (std::list<Exp*>::iterator i=mEqualities.begin();
+                i != mEqualities.end(); i++) {
+        is = (*i)->isUFSArg(tvar,false);
+        if (is)  return is;
+    }
+ 
+     // Iterate over the inequality expressions.
+    for (std::list<Exp*>::iterator i=mInequalities.begin();
+                i != mInequalities.end(); i++) {
+        is = (*i)->isUFSArg(tvar, false);
+        if (is)  return is;
+    }
+
+  return is;
+}
+
+// Uses isl library to project out a tuple variable from an affine set
+std::string ISL_project_out (std::string relstr, unsigned pos)
+{
+  // Get an isl context
+  isl_ctx *ctx;
+  ctx = isl_ctx_alloc();
+  
+  // Get an isl printer and associate to an isl context
+  isl_printer * ip = NULL;
+  ip = isl_printer_to_str(ctx);
+  
+  // load Relation r into ISL map
+  isl_set* iset = NULL;
+  iset = isl_set_read_from_str(ctx, relstr.c_str());
+
+  // Project out tuple variable No. pos
+  iset = isl_set_project_out(iset, isl_dim_out, pos, 1);
+
+  // get string back from ISL map
+  char * cstr;
+  isl_printer_set_output_format(ip , ISL_FORMAT_ISL);
+  isl_printer_print_set(ip ,iset);
+  cstr=isl_printer_get_str(ip);
+  std::string stringFromISL = cstr;
+  
+  // clean-up
+  isl_printer_flush(ip);
+  isl_printer_free(ip);
+  free(cstr);
+  isl_set_free(iset);
+  iset= NULL;
+  isl_ctx_free(ctx); 
+
+  return stringFromISL;
+}
+
+/* Mahdi Soltan Mohammadi
+     Projects out tuple varrable No. tvar
+     tvar is calculated based on total ariety (in+out) starting from 0.
+     Consequently, to project out jp from R: tvar = 2
+     R = {[i,j,k] -> [ip,jp,kp] : ...}
+*/
+void Conjunction::project_out(int tvar)
+{
+    /////////////////////
+    // (step 0) Group together all equality expressions that 
+    // are parts of the same UFCallTerm, IOW i=f(k)[0] and 
+    // j=f(k)[1] should become (i,j) = f(k).
+    // FIXME: what about inequalities?
+
+    Conjunction* selfcopy = new Conjunction(*this);
+    TupleDecl origTupleDecl = selfcopy->getTupleDecl(); // for (step 3)
+    selfcopy->groupIndexedUFCalls();
```

Comment:
This is needed for normalization, but I don't think you need it here.  Take it out and simplify the comment for step 0.  You probably just need origTupleDecl.